### PR TITLE
Waste fewer layers when copying resources

### PR DESCRIPTION
**Pull request by dflemstr**
Before this PR, the `docker-maven-plugin` would create one layer per copied file in a `<resources>` block. This now collates files that end up in the same directory into one command. This also uses `COPY` instead of `ADD` for cleaner semantics.

**tl;dr** before:

    COPY a.txt /foo/a.txt
    COPY b.txt /foo/b.txt
    COPY c.txt /foo/a.txt

after:

    COPY a.txt b.txt c.txt /foo/

### CODE REVIEW

**Comment by rculbertson** [comment]
Good discussion. I'm inclined to merge if we have concrete use cases which benefit from batching the copy operations, but haven't identified any which would suffer from this.

@davidxia?

**Comment by dflemstr** [comment]
The use case I described above is not enough?

**Comment by rculbertson** [comment]
Sorry, I worded that poorly. I meant to say that the only reason not to merge is that we lose the per-file caching, but I agree with your point that it's not that important. Since we don't have a use case that would suffer from losing per-file caching, I'm inclined to merge.

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 1]
```diff
@@ -531,9 +548,9 @@ private void createDockerFile(String directory, List<String> filesToAdd) throws
     Files.write(Paths.get(directory, "Dockerfile"), commands, UTF_8);
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: @dflemstr Does this line compile for you? Paths.get() expects a Paths object or Strings but directory is now a Path object.

**Comment by davidxia** [comment]
I'm inclined to merge as well. I think we also need to update `BuildMojoTest`'s `GENERATED_DOCKERFILE` to have `COPY ...` instead of `ADD`.

**Comment by dflemstr** [comment]
Currently, for our project that has this problem, we have started using the `.tar` solution described above (using a manual Dockerfile) and it turns out to be really nice. We might want to consider that solution before merging this PR

File: src/main/java/com/spotify/docker/BuildMojo.java (hunk 1 above)
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: I think this should do it.

```
    Files.createDirectories(directory);
    Files.write(directory.resolve("Dockerfile"), commands, UTF_8);
```

**Comment by rculbertson** [comment]
Closing PR since it seems it's not needed. Feel free to reopen if we want to revisit this.

**Pull request closed without merging**